source/agora/common/BitMask.d

```diff
@@ -0,0 +1,230 @@
+/*******************************************************************************
+
+    This is an implementation of a bitmask that once initialized will have a
+    fixed number of bits for holding `true` / `false` (`1` / `0`) values.
+    It allocates the required number of `ubytes` in the constructor and does not
+    allow reading or writing to bits beyond the fixed count which is set during
+    construction.
+    This type is created for use as the validators signing bitmask and any
+    changes should ensure that it does not compromise that use.
+
+*******************************************************************************/
+
+module agora.common.BitMask;
+
+public struct BitMask
+{
+    import std.algorithm;
+    import std.conv;
+    import std.range;
+
+    @safe:
+
+    /// Pretty-print this value (to save bandwidth the shorter list is created)
+    /// `3/5 !(0,2)` can be read as 3 out of 5 with indices 0 and 2 not set (i.e. `01011`)
+    /// `2/5 (0,2)` can be read as 2 out of 5 with indices 0 and 2 set (i.e. `10100`)
+    /// If half are set we list the set bit indices e.g. `3/6 (0,1,2)` = `111000`
+    public void toString (scope void delegate (scope const char[]) @safe sink) const
+    {
+        import std.format;
+        bool list_set = this.setCount <= this.length / 2;
+        formattedWrite(sink, "%d/%d %s", this.setCount, this.length,
```

Comment:
```suggestion
        formattedWrite(sink, "%d/%d %(%s, %)", this.setCount, this.length,
```

And then just give it a range.

Reply:
What about the `!`?

Reply:
I say, let's just list all missing validators, even if the list is longer.
This is for human consumption, I don't think we should use this for Vibe.d serialization. I'll have a look at what is available.

Reply:
But the idea is to reduce the bandwidth by always using the shortest list.

Reply:
What bandwidth ?

Reply:
If we have 1000 validators and we list 999 indices yet to sign we use more bandwidth than listing the first one that signs.